### DISCUSSION TITLE
Why frameworks?

### DISCUSSION THREAD

**Discussion by veganaiZe** [ykdojo/defaang#18] [Tech]
Why already worrying about frameworks?  Why not just go vanilla and build up from there?  :icecream:

**Comment by ykdojo** [comment]
I'm just familiar with Next.js + Tailwind, but I have nothing against vanilla. Open to ideas

**Reply by veganaiZe** [reply]
Maybe starting with pure HTML & CSS (possibly using Tailwind) and see how far it can go _--without losing control, of course--_ like that?

HTML, followed by CSS, are the quickest for a browser to render.  To my knowledge they can be processed in parallel, across multiple threads.

I don't know about anyone else, but I'm about over websites which render & react more slowly than those of last century.  Most of the trade-offs seem to comfort the developer, over the end user, not to mention the potential pollution caused by millions of such bloated apps?

**Reply by ykdojo** [reply]
The only challenge with pure HTML, CSS & JS is integrating with Supabase or any other backend that we decide to use.

Assuming we use either [Supabase or Firebase](https://github.com/ykdojo/defaang/discussions/23) for it, we'll need to load data from there and render it.

I believe with Next.js, we can render them on the server side, so that should work well.

I guess another alternative is to build a monolithic web app with something like Rails or Django, but I'm not quite interested in doing that for this particular project.

**Reply by veganaiZe** [reply]
[Supabase supports vanilla access](https://github.com/supabase/examples/tree/main/supabase-js-v1/auth/javascript-auth).

Server side rendering definitely sounds like the right approach, at least initially.  But why is a 3rd-party framework necessary to do so?  It seems like a roughly equal amount of vanilla could also do the trick (server-side via Node)?  The pages could simply be served as regular pre-rendered HTML/CSS/JS pages...

I wrote [a little proof-of-concept](https://gist.github.com/veganaize/fc3b9aa393ca688a284c54caf43a3fc3), in Node.js, demonstrating how vanilla (no Express) can very efficiently serve pages and all of the code fits within a single screen of an old school (80x24) terminal.

_Thanks for entertaining these thoughts!_

**Reply by ykdojo** [reply]
So Node, Rails, Django, or anything else - I'm just not excited about the idea of working with any of them, tbh.

Again, it's a combination of the fact that I'm more familiar with Next.js, and also the fact that I'm not excited about the idea of making and managing our own backend.

I totally see your point, though.

**Reply by veganaiZe** [reply]
Gotcha.  Sorry, I'd thought you were already using node on the backend.

**Comment by subhoghoshX** [comment]
I'm personally against vanilla. It's like being a masochist. If we're too much into performance or JS file size, we can go for something like Astro. But I think NextJS holds a nice balance between performance and developer experience :)

**Reply by veganaiZe** [reply]
By _vanilla_ I mean as little 3rd-party frameworks as possible, within reason, unless empirically justified.  Can you offer any substantial evidence as to why vanilla is _masochistic_?  How much _(single-threaded)_ Javascript does a site like this really need, anyway?

Thank you for your thoughts.

**Comment by thepushkaraj** [comment]
I think vanilla JS is good for small projects, and it's recommended in that situation. 

But I think it's going to be complex project with lots of pages and content, so it makes our life easier as a developer, that's why frameworks are recommended in this situation. 

What do you think @ykdojo ?

**Reply by ykdojo** [reply]
I think I've explained most of my thoughts here: https://github.com/ykdojo/defaang/discussions/18#discussioncomment-3377920

**Reply by veganaiZe** [reply]
I don't see what's going to be so complex about this project.  Piles of frameworks feel like complexity to me.  But I know I'm the extreme minority :neckbeard:  and that most people are used to learning frameworks instead of markup & programming languages.
